lib/nrf_modem_lib/trace_medium.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/logging/log.h>
+#include <nrf_modem.h>
+#include <modem/trace_medium.h>
+
+LOG_MODULE_REGISTER(trace_medium, CONFIG_NRF_MODEM_LIB_LOG_LEVEL);
+
+static struct trace_medium_entry *medium = NULL;
+
+bool trace_medium_init(void)
+{
+	int err;
+
+	/* Attempt to reinitialize medium */
+	if (medium != NULL) {
+		err = medium->init();
+		if (err == 0 || err == EALREADY) {
+			return true;
+		} else {
+			LOG_ERR("Failed to reinit trace medium %s, err %d", medium->name, err);
+			medium = NULL;
+		}
+	}
+
+	/* Find and attempt to initialize a medium */
+	STRUCT_SECTION_FOREACH(trace_medium_entry, m) {
+		if (m->init == NULL || m->deinit == NULL || m->write == NULL) {
+			continue;
+		}
+
+		err = m->init();
+		if (!err) {
+			medium = m;
+			LOG_INF("Modem trace medium %s initialized", medium->name);
+			return true;
+		} else {
+			LOG_ERR("Failed to init trace medium %s, err %d", m->name, err);
+		}
+	}
+
+	return false;
+}
+
+bool trace_medium_deinit(void)
+{
+	int err;
+
+	if (medium == NULL) {
+		return true;
+	}
+
+	err = medium->deinit();
+	if (!err) {
+		LOG_INF("Modem trace medium %s deinitialized", medium->name);
+		medium = NULL;
+		return true;
+	} else {
+		LOG_ERR("Failed to deinit trace medium %s, err %d", medium->name, err);
+	}
+
+	return false;
+}
+
+void trace_medium_write(const uint8_t *data, uint32_t len)
```

Comment:
We should log an error if we call `trace_medium_write` without a valid trace medium